### PR TITLE
Moving Optimism bin to Optimism Cli

### LINKED ISSUE #9408: add optimism binary in optimism-cli

### PR DESCRIPTION
**Pull request by loocapro**
Closes https://github.com/paradigmxyz/reth/issues/9408

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
great!
a few more nits, then this is g2g
I'll need to go over a few things before we can merge this, because we need to update a few scripts that build the op-reth binary

File: crates/optimism/bin/src/main.rs [hunk 1]
```diff
@@ -10,11 +9,8 @@ use std::sync::Arc;
 #[global_allocator]
 static ALLOC: tikv_jemallocator::Jemalloc = tikv_jemallocator::Jemalloc;
 
-#[cfg(not(feature = "optimism"))]
-compile_error!("Cannot build the `op-reth` binary with the `optimism` feature flag disabled. Did you mean to build `reth`?");
-
-#[cfg(feature = "optimism")]
```
Comment (emhane) [inline crates/optimism/bin/src/main.rs:16]: lfg

File: crates/optimism/cli/src/lib.rs [hunk 2]
```diff
@@ -86,3 +93,90 @@ impl Cli {
...
+    /// }
+    ///
+    /// Cli::parse()
+    ///     .run(|builder, my_args: MyArgs| async move {
+    ///         // launch the node
+    ///
+    ///         Ok(())
+    ///     })
+    ///     .unwrap();
+    /// ````
```
Comment (emhane) [inline crates/optimism/cli/src/lib.rs:139]: maybe out of scope, but I would prefer these to be executable examples. easy to keep up to date with potential type name changes then using `cargo test --doc`.

**Review by emhane — CHANGES_REQUESTED** [review]
this pr should also remove the optimism commands from the `reth` binary 

https://github.com/loocapro/reth/blob/39b0cf61ccc646beee64b5de6eb36c42ee252695/bin/reth/src/cli/mod.rs#L157-L162

and remove the whole `optimism` feature from `bin/reth`

https://github.com/loocapro/reth/blob/39b0cf61ccc646beee64b5de6eb36c42ee252695/bin/reth/Cargo.toml#L135-L146

File: crates/optimism/cli/src/lib.rs (hunk 2 above)
Comment (mattsse) [reply crates/optimism/cli/src/lib.rs:139]: we can finally do this now actually

**Review by mattsse — COMMENTED** [review]
tysm!

**Review by mattsse — CHANGES_REQUESTED** [review]
just checked, we need to do a few more things before we can finalize this

## TODO
* remove all `-optimism-` crates from bin/reth
* remove all usages of cfg(optimism) and only keep the ethereum code, like https://github.com/paradigmxyz/reth/blob/main/bin/reth/src/commands/debug_cmd/build_block.rs#L249-L253

**Review by mattsse — CHANGES_REQUESTED** [review]

File: bin/reth/src/commands/debug_cmd/build_block.rs
Comment (emhane) [inline bin/reth/src/commands/debug_cmd/build_block.rs:1]: the code that has been deleted from `bin/reth`, needs to move to `reth-optimism-cli`. so this requires having a debug command in `reth-optimism-cli`.

**Review by emhane — DISMISSED** [review]

File: crates/optimism/cli/src/lib.rs (hunk 2 above)
Comment (emhane) [reply crates/optimism/cli/src/lib.rs:137]: either it's in scope and this pr fixes it, or open an issue pls

**Comment by emhane** [comment]
how's it going with this? need some help? I have a PR that's a dependent on this @loocapro

**Comment by loocapro** [comment]
> how's it going with this? need some help? I have a PR that's a dependent on this @loocapro

Hi @emhane, since i removed some cfg flags, I am running in some workspace dependency issue here:

https://github.com/paradigmxyz/reth/actions/runs/9909282475/job/27376994891?pr=9439

Any advice?

File: crates/optimism/bin/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+homepage.workspace = true
+repository.workspace = true
+exclude.workspace = true
+
+[dependencies]
+reth-node-builder.workspace = true
+reth-cli-util.workspace = true
+reth-optimism-cli.workspace = true
+reth-tracing.workspace = true
+reth-node-optimism.workspace = true
```
Comment (emhane) [inline crates/optimism/bin/Cargo.toml:16]: in response to https://github.com/paradigmxyz/reth/pull/9439#issuecomment-2227409172:

you will have to add an optimism feature to this manifest, and then look through these deps, and see which has an optimism feature...so they can be included, just like for the other optimism crates - eventually we hope to get rid off this!

**Comment by emhane** [comment]
how's it going @loocapro ?

**Comment by mattsse** [comment]
only thing missing here is getting ahold of the feature snafu @joshieDo and I are getting this over the line

**Review by mattsse — APPROVED** [review]

**Comment by loocapro** [comment]
> release builds work fine: https://github.com/paradigmxyz/reth/actions/runs/10144116336/job/28046941974?pr=9439

Thanks @joshieDo for taking this on!! Amazing work!!

**Pull request merged**